Sources/BraintreeCoreSwift/BTHTTP.swift

```diff
@@ -0,0 +1,542 @@
+                    self.handleRequestCompletion(data: cachedResponse.data, request: nil, shouldCache: false, response: cachedResponse.response, error: nil, completion: completion)
+                } else {
+                    let task: URLSessionTask = self.session.dataTask(with: request) { [weak self] data, response, error in
+                        self?.handleRequestCompletion(data: data, request: request, shouldCache: true, response: response, error: error, completion: completion)
+                    }
+
+                    task.resume()
+                }
+            }
+        }
+    }
+
+    func httpRequest(
+        method: String,
+        path: String,
+        parameters: NSDictionary? = [:],
+        completion: RequestCompletion?
+    ) {
+        createRequest(method: method, path: path, parameters: parameters) { request, error in
+            guard let request = request else {
+                self.handleRequestCompletion(data: nil, request: nil, shouldCache: false, response: nil, error: error, completion: completion)
+                return
+            }
+
+            let task: URLSessionTask = self.session.dataTask(with: request) { [weak self] data, response, error in
+                self?.handleRequestCompletion(data: data, request: request, shouldCache: false, response: response, error: error, completion: completion)
+            }
+
+            task.resume()
+        }
+    }
+
+    func createRequest(
+        method: String,
+        path: String,
+        parameters: NSDictionary? = [:],
+        completion: @escaping (URLRequest?, Error?) -> Void
+    ) {
+        let hasHTTPPrefix: Bool = path.hasPrefix("http")
+        let baseURLString: String = baseURL.absoluteString
+        var errorUserInfo: [String: Any] = [:]
+
+        if hasHTTPPrefix && (baseURLString.isEmpty || baseURLString == "") {
+            errorUserInfo["method"] = method
+            errorUserInfo["path"] = path
+            errorUserInfo["parameters"] = parameters
+
+            let error = Self.constructError(code: .missingBaseURL, userInfo: errorUserInfo)
+
+            completion(nil, error)
+            return
+        }
+        
+        let fullPathURL: URL?
+        let isDataURL: Bool = baseURL.scheme == "data"
+
+        if !isDataURL {
+            fullPathURL = hasHTTPPrefix ? URL(string: path) : baseURL.appendingPathComponent(path)
+        } else {
+            fullPathURL = baseURL
+        }
+
+        let mutableParameters: NSMutableDictionary = NSMutableDictionary(dictionary: parameters ?? [:])
+
+        if case .authorizationFingerprint(let fingerprint) = clientAuthorization {
+            mutableParameters["authorization_fingerprint"] = fingerprint
+        }
+
+        guard let fullPathURL = fullPathURL else {
+            // baseURL can be non-nil (e.g. an empty string) and still return nil for -URLByAppendingPathComponent:
+            // causing a crash when NSURLComponents.componentsWithString is called with nil.
```

Comment:
```suggestion
            // baseURL can be non-nil (e.g. an empty string) and still return nil for appendingPathComponent(_:)
            // causing a crash when URLComponents(string:_) is called with nil.
```

Update to swift method names